packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts

```diff
@@ -0,0 +1,305 @@
+   */
+  private resolveReferences(ref: any): any {
+    if (Array.isArray(ref)) {
+      return ref.map(r => this.resolveReferences(r));
+    } else if (typeof ref === 'object') {
+      return this.resolveJsonObject(ref);
+    }
+    const resolvable = Tokenization.reverseString(ref);
+    if (resolvable.length === 1 && Reference.isReference(resolvable.firstToken)) {
+      return `(${resolvable.firstToken.target.node.path}.${resolvable.firstToken.displayName})`;
+    } else {
+      const resolvedTokens = resolvable.mapTokens({
+        mapToken: (r: IResolvable) => {
+          if (Reference.isReference(r)) {
+            return `(${r.target.node.path}.${r.displayName})`;
+          }
+          const resolved = Stack.of(this).resolve(r);
+          if (typeof resolved === 'object' && resolved.hasOwnProperty('Ref')) {
+            switch (resolved.Ref) {
+              case 'AWS::AccountId':
+                return '(ACCOUNT)';
+              case 'AWS::Partition':
+                return '(PARTITION)';
+              case 'AWS::Region':
+                return '(REGION)';
+              default:
+                return r;
+            }
+          }
+          return r;
+        },
+      });
+      return resolvedTokens.join(new StringConcat());
+    }
+  }
+
+  private resolveJsonObject(statement: { [key: string]: any }): any {
+    const newStatement = statement;
+    for (const [key, value] of Object.entries(statement)) {
+      newStatement[key] = this.resolveReferences(value);
+    }
+    return newStatement;
+  }
+}
+
+/**
+ * Options for generating the role policy report
+ */
+interface RoleReportOptions {
+  /**
+   * The name of the IAM role.
+   *
+   * If this is not provided the role will be assumed
+   * to be missing.
+   *
+   * @default 'missing role'
+   */
+  readonly roleName?: string;
+
+  /**
+   * A list of IAM Policy Statements
+   */
+  readonly policyStatements: string[];
+
+  /**
+   * A list of IAM Managed Policy ARNs
+   */
+  readonly managedPolicies: string[];
+
+  /**
+   * The trust policy for the IAM Role.
+   *
+   * @default - no trust policy.
+   */
+  readonly assumeRolePolicy?: string;
+}
+
+/**
+ * A construct that is responsible for generating an IAM policy Report
+ * for all IAM roles that are created as part of the CDK application.
+ *
+ * The report will contain the following information for each IAM Role in the app:
+ *
+ * 1. Is the role "missing" (not provided in the customizeRoles.usePrecreatedRoles)?
+ * 2. The AssumeRole Policy (AKA Trust Policy)
+ * 3. Any "Identity" policies (i.e. policies attached to the role)
+ * 4. Any Managed policies
+ */
+class PolicySynthesizer extends Construct {
+  private readonly roleReport: { [roleName: string]: RoleReportOptions } = {};
+  constructor(scope: Construct) {
+    super(scope, POLICY_SYNTHESIZER_ID);
+
+    // attach a custom synthesis during validation, which happens
+    // _after_ all aspects are run and all references are resolved
```

Comment:
Why is this important?

If the goal is to not update the report unless synthesis completes successfully, I think that's more appropriately a part of the synthesizer protocol (to be able to explicitly request this behavior) than trying to employ your knowledge of the CDK lifecycle to do these order-dependent things.

In general, order-dependent coding is scary and dangerous (to me).

Reply:
yeah I don't think it's actually important that this be _attached_ at a specific point since really all we care about is that the synthesizer runs as part of synthesis, which it will regardless of when we attach it. I've removed the `addValidation` piece.